### ISSUE TITLE
RegularizedLDLT gets stuck in an infinite loop for some problems

### ISSUE THREAD

**Issue by JaciBrunning** [SleipnirGroup/Sleipnir#74]
The following loop gets stuck for some problems, resulting in an infinite loop that ignores timeout and iteration limit conditions:
https://github.com/SleipnirGroup/Sleipnir/blob/main/src/optimization/RegularizedLDLT.hpp#L95-L119

The following code is sufficient to generate this error, presumably due to the addition of the height limit constraint. 
```cpp
  sleipnir::OptimizationProblem problem;

  double total_time = 4;
  auto dt = total_time / _N;

  auto elevator = problem.DecisionVariable(2, _N + 1);
  auto elevator_accel = problem.DecisionVariable(1, _N);

  auto arm = problem.DecisionVariable(2, _N + 1);
  auto arm_accel = problem.DecisionVariable(1, _N);

  for (size_t k = 0; k < _N; k++) {
    problem.SubjectTo(elevator(0, k+1) == elevator(0, k) + elevator(1, k) * dt);
    problem.SubjectTo(elevator(1, k+1) == elevator(1, k) + elevator_accel(0, k) * dt);

    problem.SubjectTo(arm(0, k+1) == arm(0, k) + arm(1, k) * dt);
    problem.SubjectTo(arm(1, k+1) == arm(1, k) + arm_accel(0, k) * dt);
  }

  // Start and End Conditions
  problem.SubjectTo(elevator.Col(0) == Eigen::Vector2d({ start.height.value(), 0.0 }));
  problem.SubjectTo(elevator.Col(_N) == Eigen::Vector2d({ end.height.value(), 0.0 }));

  problem.SubjectTo(arm.Col(0) == Eigen::Vector2d({ start.angle.value(), 0.0 }));
  problem.SubjectTo(arm.Col(_N) == Eigen::Vector2d({ end.angle.value(), 0.0 }));

  // Velocity and Acceleration Limits
  problem.SubjectTo(-_constraints.max_velocity_elevator.value() <= elevator.Row(1));
  problem.SubjectTo(elevator.Row(1) <= _constraints.max_velocity_elevator.value());

  problem.SubjectTo(-_constraints.max_acceleration_elevator.value() <= elevator_accel);
  problem.SubjectTo(elevator_accel <= _constraints.max_acceleration_elevator.value());

  problem.SubjectTo(-_constraints.max_velocity_arm.value() <= arm.Row(1));
  problem.SubjectTo(arm.Row(1) <= _constraints.max_velocity_arm.value());

  problem.SubjectTo(-_constraints.max_acceleration_arm.value() <= arm_accel);
  problem.SubjectTo(arm_accel <= _constraints.max_acceleration_arm.value());

  // Height Limit
  auto y = elevator.Row(0) + _config.arm.armLength.value() * sleipnir::sin(arm.Row(0));
  problem.SubjectTo(y <= _constraints.max_height_end_effector.value());

  // Cost Function
  sleipnir::Variable costTotal = 0.0;
  for (int k = 0; k < _N + 1; k++) {
    costTotal += sleipnir::pow(end.height.value() - elevator(0, k), 2) + sleipnir::pow(end.angle.value() - arm(0, k), 2);
  }
  problem.Minimize(costTotal);
```

Where:
- `_config.arm.armLength = 1m`
- `_constraints.max_velocity_elevator = 1m/s`
- `_constraints.max_acceleration_elevator = 2m/s/s`
- `_constraints.max_velocity_arm = 360deg/s`
- `_constraints.max_acceleration_arm = 720deg/s/s`
- `_constraints.max_height_end_effector = 1.8m`
- `start.height = 1m`
- `start.angle = 0deg`
- `end.height = 1.25m`
- `end.angle = 180deg`

**Comment by calcmogul** [comment]
tl;dr The equality constraint Jacobian is rank-deficient (duplicate constraints) and has positive eigenvalues, and that's breaking the solver. This could be due to your poor initial guess of zeros for all your decision variables. Try initializing them with a simple trajectory (line segments would suffice).

Here's some debugging notes.

This problem has 4804 decision variables and 3208 equality constraints. The Newton-KKT step solves a system with the following left-hand side.
```
[H + AᵢᵀΣAᵢ  Aₑᵀ]
[    Aₑ       0 ]
```
The inertia of this matrix (the number of positive, negative, and zero eigenvalues) must be (4804, 3208, 0) to obtain a solution, which means the Hessian must be positive definite and the equality constraint Jacobian must be negative definite.

If it's not, a regularization procedure is performed that increases δ and γ in the following system.
```
lhs = [H + AᵢᵀΣAᵢ + δI   Aₑᵀ]
      [       Aₑ        −γI ]
```
The inertia of Aₑ should be (0, 3208, 0) but is actually (2591, 612, 5), so that's not a good start.

The inertia of lhs is initially (4802, 3210, 0) (ideal inertia is (4804, 3208, 0)). γ stays at zero because lhs isn't rank-deficient. δ is increased by powers of 10 to try to fix the inertia. It eventually converges to: 
```
idealInertia=(4804, 3208, 0)
inertia=(4801, 3210, 1)
```
δ increases in powers of ten until it hits infinity, then the regularizer is stuck forever.

ACTION: We should abort the solve in this case with a diagnostic. I'm not sure yet whether we should also abort if the equality constraint Jacobian has positive eigenvalues.

My guess is the equality constraint being rank deficient and not at least negative semidefinite is what's causing the problem. It doesn't cause negative curvature in all cases, but that nonlinear inequality constraint seems to push it over the edge.

Try providing an initial guess for the solution by initializing the decision variables to nonzero values.

**Comment by calcmogul** [comment]
Looks like in addition to what we're already doing, ipopt reverts to a feasibility restoration phase if no value of δ makes the Newton-KKT nonsingular.
![Screenshot_20230130_122509](https://user-images.githubusercontent.com/2748555/215587267-8a3f47ba-274d-4938-a36f-34c5c5b2f64e.png)

**Comment by calcmogul** [comment]
I modified the Newton-KKT solve to set the step length to zero if the regularization procedure fails so second-order corrections can attempt to restore feasibility. It seems to converge now. I need to add a test for it.

**Comment by calcmogul** [comment]
If someone could run the ArmOnElevator unit test on Windows with a debugger to get a backtrace for its segfault, that would be great. I don't have a Windows dev environment.

**Comment by calcmogul** [comment]
The PR uses second-order corrections to improve feasibility, but that still uses the potentially ill-conditioned Newton-KKT system. We should probably look at implementing a proper feasibility restoration mode that tries to minimize constraint violation.
